src/hrflow_connectors/connectors/lever/warehouse.py

```diff
@@ -0,0 +1,396 @@
+                    for opportunity in opportunities:
+                        opportunity_id = opportunity["id"]
+                        profile_response = requests.get(
+                            f"{LEVER_OPPORTUNITIES_ENDPOINT}/{opportunity_id}/resumes",
+                            headers=headers,
+                        )
+                        if profile_response.status_code == 200:
+                            profile_data = profile_response.json()
+                            opportunity["profile"] = profile_data["data"]
+                            yield opportunity
+                        else:
+                            adapter.error(
+                                "Failed to retrieve profiles for opportunity"
+                                f" {opportunity_id}. Status code:"
+                                f" {profile_response.status_code}, Response:"
+                                f" {profile_response.text}"
+                            )
+
+                    # After fetching a page of results
+                    if response.json().get("hasNext", False):
+                        offset = response.json().get("next", None)
+                    else:
+                        break  # No more pages, exit the loop
+                # manage rate limit
+                elif response.status_code == 429:
+                    adapter.error("Rate limit exceeded. Retrying after 1 minute.")
+                    time.sleep(60)
+                else:
+                    adapter.error(
+                        "Failed to retrieve opportunities from Lever. Status code:"
+                        f" {response.status_code}, Response: {response.text}"
+                    )
+                    break
+            except requests.exceptions.RequestException as e:
+                if "401" in str(e):
+                    adapter.error(
+                        "Error: an exception occurred while fetching opportunities {e}"
+                    )
+                    token, refresh_token = get_or_refresh_tokens(
+                        parameters.client_id,
+                        parameters.client_secret,
+                        "refresh_token",
+                        refresh_token=refresh_token,
+                    )
+                    if not token:
+                        raise Exception("Failed to refresh token.")
+                else:
+                    raise Exception(f"Request failed with error: {e}")
+        else:
+            raise Exception("Failed to obtain initial access token.")
+
+
+def write(
+    adapter: LoggerAdapter,
+    parameters: WriteProfileParameters,
+    profiles: t.Iterable[t.Dict],
+) -> t.List[t.Dict]:
+    token, refresh_token = get_or_refresh_tokens(
+        parameters.client_id,
+        parameters.client_secret,
+        "authorization_code",
+        parameters.authorization_code,
+    )
+    failed_profiles = []
+    for profile in profiles:
```

Comment:
Le problème ici c'est qu'en faisant comme ca i.e. `for profile in profiles` ca te force à l'interieur du sous block `status_code = 401` à dupliquer le logiue du post et une partie de la gestion d'erreur. 

En général éviter la duplication c'est plus important que le style du code. Tu a au moins deux alternatives

```
i = 0 
while i < len(profiles):
     profile = profiles[i]
     response = request...
     if success:
           ....
           i += 1 
     elif status_code = 401 : 
          # juste le code du refresh_token et simplement en faisant pas i + =1 tu reviens au debut de la boucle while et tu retry le même profile 
```

Ou alors tu mets  tu refactor la partie que tu as été obligé de dupliquer dans une fonction. Et mnt tu as deux appels mais le code est à un seul endroit

Reply:
parfait! j'ai utilisé le while